### PR TITLE
phase 0: <algorithm> sort-family bindings + top_n_by

### PR DESCRIPTION
**Pull request by borisbat**
## Summary

Phase 0 of the linq_fold splice project — foundational `<algorithm>` sort-family bindings that unblock the BufferTopN emit mode in PR B. Adds C++ bindings for `partial_sort`, `nth_element`, `make_heap`, `push_heap`, `pop_heap` across both the typed (19 workhorse types) and generic byte-pointer paths, plus daslib wrappers, dispatcher macros, top_n/top_n_by in linq, tests, benchmarks, docs, and an expanded tutorial.

The existing `__builtin_sort` surface had no partial-sort, nth-element, or heap-op equivalents. `daslib/linq.order_by(...) |> take(N)` falls through to full O(M log M) sort then slice — 60× slower than SQLite's partial-sort + LIMIT at 100K rows. Phase 0 closes the C++ binding gap so PR B can splice `[where][select]* |> order_by(...) |> {take(N), first}` into a single fused loop at O(M log N).

Plan: `~/.claude/plans/merry-dazzling-curry.md`. Project context: `benchmarks/sql/LINQ.md`.

### What's added

**C++ algorithm layer** (`src/builtin/das_qsort_r.h`, ~150 LOC)
- `das_nth_element_r` — introselect (median-of-three quickselect with smoothsort fallback on depth blowup)
- `das_partial_sort_r` — `nth_element` + smoothsort the first N
- `das_make_heap_r` / `das_push_heap_r` / `das_pop_heap_r` — binary max-heap on contiguous byte buffer
- Same inline-comparator template style as the existing musl-derived `das_qsort_r`
- Without these, partial_sort/nth_element/heap ops on user-defined struct types would have no generic path (std::sort templates on iterator type, not void*+width)

**C++ bindings** (`include/daScript/simulate/aot.h`, `aot_builtin.h`, `module_builtin_runtime_sort.cpp`)
- Typed cblock templates (`scblk_*` structs + JIT-block fast path) for all 5 ops
- ~380 typed registrations (5 ops × 19 numeric+vector workhorse types × 4 variants)
- ~30 generic any-path registrations (5 ops × 6 variants each) routed through `das_*_r`
- All wrapped in `STAMP_NUMERIC_OPS` / `STAMP_VECTOR_OPS` / `REGISTER_ANY_PATH[_N]` macros for diff readability

**Daslib wrappers** (`daslib/sort_boost.das`)
- User-facing `partial_sort` / `nth_element` / `make_heap` / `push_heap` / `pop_heap` for `array<T>` (with and without custom comparator block)
- `q*` dispatch macros (`qpartial_sort`, `qnth_element`, `qmake_heap`, `qpush_heap`, `qpop_heap`) sharing `build_dispatch_call` helper; mirror the existing `qsort` macro pattern for handled vector types via `temp_array(...)`

**Top-N selection in linq** (`daslib/linq.das`)
- `top_n(arr_or_iter, n)` and `top_n_by(arr_or_iter, n, key)` — returns N smallest sorted ascending
- Array source: `partial_sort` + resize (one allocation)
- Iterator source: bounded heap of size N during scan, at most N+1 elements ever resident; O(M log N) total
- Empty / N=0 / N>length all handled

### Tests

- **C++ tests** (`tests-cpp/small/test_sort_family.cpp`) — 7 TEST_CASEs, 110 assertions covering all templates including the previously-untested `das_qsort_r` baseline (`ctest` 33/33 pass)
- **Daslang tests** (`tests/linq/test_linq_sorting.das`) — 4 new `[test]` blocks (25 subtests) covering partial_sort/nth_element/top_n_by/heap_ops; default + custom-comparator + key-fn forms; struct payload preservation; edge cases (N=0, N=length, N>length, k out-of-range); user-struct any-path. Full suite passes 53/53.
- **Tutorial smoke** — `tutorials/language/33_algorithm.das` runs end-to-end with correct output for all 12 demo sections (4 new + 8 existing)

### Benchmarks

New files (numbers to capture at review time):
- `benchmarks/sort/sort.das` — baseline sort (was previously missing)
- `benchmarks/sort/partial_sort.das` — partial_sort vs full-sort-then-take, sweep N=1/10/100/1000
- `benchmarks/sort/nth_element.das` — nth_element vs full-sort-then-index for head/median/tail positions
- `benchmarks/sort/heap_ops.das` — make_heap+pop_heap build-extract loop; bounded-N streaming insert
- `benchmarks/sql/top_n_by_bench.das` — 4-way m1/m3/m3f_old/m3f columns + new m3_topn_array/m3_topn_iter columns

### Documentation

- `//!` doc-comments on every new public function/macro
- `doc/reflections/das2rst.das` — new groups: `Partial sorting`, `Heap operations` (under `document_module_sort_boost`); `Top-N selection` (under `document_module_linq`)
- Stdlib docs regenerate clean (no Uncategorized, no stubs)
- `tutorials/language/33_algorithm.{das,rst}` expanded with 4 new sections: `sort`, `partial_sort / nth_element`, `Heap operations`, `top_n / top_n_by`

### Out of scope / known limitations

- **No dim (`[]`) overloads in the daslib wrappers** — array form only. Users with fixed-size dim arrays can call `__builtin_*` directly. Phase 0.x follow-up if needed.
- **Buffer-required emit modes in linq_fold** (BufferTopN consuming `partial_sort`, BufferDistinct, BufferGroupBy, BufferReverse, MultiSourceZip, BufferedJoin) — Phase 3+ PRs.
- **String type omitted from new bindings** — `__builtin_sort_string` has special strcmp handling; partial_sort/heap-ops on strings is not a Phase 0 use case.
- **`benchmarks/sql/sort_take.das` pre-existing compile failure** on master (`each(arr)` unsafe inside `_fold` macro) — unrelated; should be a separate fix.

## Test plan

- [x] `mcp__daslang__compile_check` and `lint` clean on all modified `.das` files
- [x] `mcp__daslang__format_file` confirms all `.das` files already formatted
- [x] Full `dastest tests/` — 8366/8372 pass (6 skipped, 0 failed, 0 errors)
- [x] `ctest` — 33/33 pass including new `test_sort_family.cpp` (7 TEST_CASEs, 110 assertions)
- [x] `tests/linq/test_linq_sorting.das` — 53/53 pass via MCP `run_test`
- [x] `tutorials/language/33_algorithm.das` runs end-to-end, exit 0
- [x] `test_aot` build clean; AOT dastest 7755/7761 pass (6 skipped, 0 failed/errored)
- [x] Clean `sphinx-build` — `build succeeded.` with zero warnings/errors
- [ ] Benchmark numbers — captured at review time

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: daslib/sort_boost.das [hunk 1]
```diff
@@ -45,3 +317,78 @@ class QsortMacro : AstCallMacro {
...
+    }
+    let blockIndex = takesN ? 2 : 1
+    macro_verify(expr.arguments[blockIndex]._type.baseType == Type.tBlock, prog, expr.at, "expecting block argument as last")
+    let srcType = expr.arguments[0]._type
+    if (!empty(srcType.dim) || srcType.baseType == Type.tArray) {
+        var call = new ExprCall(name := opName, at = expr.at)
+        for (a in expr.arguments) {
+            emplace_new(call.arguments) <| clone_expression(a)
+        }
+        return <- call
```
[the remark this answers is not in the document]
Comment (borisbat) [reply daslib/sort_boost.das]: Fixed in 0c1813de4. Dropped the `!empty(srcType.dim)` branch from `build_dispatch_call` and updated the error message to "can only {opName} array or handled vector". You are right — `qsort` works on dim because the underlying `__builtin_sort` registers dim variants; we did not add those for `partial_sort` / `nth_element` / heap ops, so accepting dim in the dispatcher would have expanded to a non-existent overload.

File: doc/source/reference/tutorials/33_algorithm.rst [hunk 2]
```diff
@@ -49,6 +53,135 @@ Custom comparators let you search in non-default order:
...
+These live in ``daslib/sort_boost``. ``partial_sort(arr, N)`` sorts only the
+first ``N`` elements ascending; the remainder is left in unspecified order:
+
+.. code-block:: das
+
+    require daslib/sort_boost
+
+    var a <- [5, 2, 8, 1, 9, 3, 7, 4, 6, 0]
+    sort_boost::partial_sort(a, 3)
+    // a[0..3] = [0, 1, 2]; a[3..] is some permutation of the rest
```
[the remark this answers is not in the document]
Comment (borisbat) [reply doc/source/reference/tutorials/33_algorithm.rst:90]: Disagreeing on this one. In daslang `0..3` is a half-open range — `range(0, 3)` is `[0, 1, 2]` (three values), and `a[0..3]` slices three elements. So "indices 0, 1, and 2" matches `a[0..3]` exactly. The comment is correct as written.

File: daslib/sort_boost.das [hunk 3]
```diff
@@ -6,15 +6,287 @@ options strict_smart_pointers = true
...
+                __builtin_partial_sort_array_any_ref_cblock(a, typeinfo sizeof(a[0]), length(a), n, cmp)
+            }
+        }
+    }
+}
+
+// ---- nth_element ----
+
+//! Place the ``n``-th smallest element at position ``n``, with all earlier
+//! elements <= and all later elements >=. Sub-O(N) on the kth element only.
```
[the remark this answers is not in the document]
Comment (borisbat) [reply daslib/sort_boost.das]: Fixed in 0c1813de4 — rewrote the doc-comment to say "O(N) average on the kth element only".

File: tests-cpp/small/test_sort_family.cpp [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+        REQUIRE(is_max_heap(a.data(), a.size()));
+        a.push_back(2);
+        das_push_heap_r(a.data(), a.size(), sizeof(int32_t), less_cmp<int32_t>());
+        CHECK(is_max_heap(a.data(), a.size()));
+        CHECK(a[0] == 10);
+    }
+}
+
+TEST_CASE("das_pop_heap_r — move max to last position, sift-down [0..n-2]") {
+    SUBCASE("repeated pop yields sorted-ascending sequence") {
```
[the remark this answers is not in the document]
Comment (borisbat) [reply tests-cpp/small/test_sort_family.cpp]: Fixed in 0c1813de4 — renamed the SUBCASE to "repeated pop yields sorted-descending sequence" so the label matches what the body verifies.

File: benchmarks/sort/nth_element.das [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+// either side. Compared against full sort + index. nth_element scales like
+// O(M); full sort like O(M log M). For median (k = M/2) nth_element should
+// dominate at any M.
+
+def run_nth_element(b : B?; n : int; k : int) {
+    b |> run("nth_element_k{k}/{n}", n) {
+        var arr <- make_int_array(n)
+        sort_boost::nth_element(arr, k)
+        if (arr[k] < 0) {
+            b->failNow()
```
[the remark this answers is not in the document]
Comment (borisbat) [reply benchmarks/sort/nth_element.das:17]: Fixed in 0c1813de4 — the benchmark fixture now uses `((i + 1) * 37) % 1000` so `arr[0] != 0` pre-sort, and a new `expected_sorted_value_at(n, i)` helper in `benchmarks/sort/_common.das` returns the exact post-sort value at position i. Each check now compares `arr[k] != expected_sorted_value_at(n, k)`, which fails on no-op runs across all k.

File: benchmarks/sort/sort.das [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+// Baseline: plain ``sort(arr, cmp)`` over int / user-struct arrays. Boris
+// flagged that sort itself had no dedicated benchmark; this fills the gap and
+// gives a reference point for the partial_sort / nth_element / heap_ops files.
+
+def run_sort_int(b : B?; n : int) {
+    b |> run("sort_int/{n}", n) {
+        var arr <- make_int_array(n)
+        sort(arr)
+        if (arr[0] > arr[1]) {
+            b->failNow()
```
[the remark this answers is not in the document]
Comment (borisbat) [reply benchmarks/sort/sort.das:15]: Fixed in 0c1813de4. Same fixture shift + helper as the nth_element fix. The new check is `arr[length(arr) - 1] != 999` (sort makes the last position the max value; pre-sort it is `(n * 37) % 1000` = `0` for n ∈ {10000, 100000}, so the check fires on no-op).

File: benchmarks/sort/partial_sort.das [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+// partial_sort(arr, N) — sort just the first N elements; remainder unspecified.
+// Compares against full sort over the same array. As N → length the two
+// converge; as N → 1 partial_sort wins by an arbitrarily large margin.
+
+def run_partial_sort_int(b : B?; n : int; k : int) {
+    b |> run("partial_sort_int_N{k}/{n}", n) {
+        var arr <- make_int_array(n)
+        sort_boost::partial_sort(arr, k)
+        if (arr[0] > arr[1]) {
+            b->failNow()
```
[the remark this answers is not in the document]
Comment (borisbat) [reply benchmarks/sort/partial_sort.das:16]: Fixed in 0c1813de4 — `arr[k - 1] != expected_sorted_value_at(n, k - 1)` for both `partial_sort` and the full-sort baseline; smoke-tested k=1, k=10, k=100, k=1000 with the shifted fixture and all distinguish pre vs post sort correctly.

File: benchmarks/sort/partial_sort.das [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+def run_partial_sort_struct(b : B?; n : int; k : int) {
+    b |> run("partial_sort_struct_N{k}/{n}", n) {
+        var arr <- make_struct_array(n)
+        sort_boost::partial_sort(arr, k) $(x, y : KeyedItem) : bool {
+            return x.key < y.key
+        }
+        if (arr[0].key > arr[1].key) {
+            b->failNow()
```
[the remark this answers is not in the document]
Comment (borisbat) [reply benchmarks/sort/partial_sort.das]: Fixed in 0c1813de4 — struct overload now checks `arr[k - 1].key != expected_sorted_value_at(n, k - 1)`. Same shifted fixture + helper as the int form.

File: benchmarks/sort/heap_ops.das [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    b |> run("build_extract/{n}", n) {
+        var arr <- make_int_array(n)
+        sort_boost::make_heap(arr)
+        while (!empty(arr)) {
+            sort_boost::pop_heap(arr)
+            arr |> resize(length(arr) - 1)
+        }
+        if (!empty(arr)) {
+            b->failNow()
+        }
```
[the remark this answers is not in the document]
Comment (borisbat) [reply benchmarks/sort/heap_ops.das:33]: Fixed in 0c1813de4 — `build_then_extract` now (a) checks `arr[0] == 999` after `make_heap` (max-at-root invariant), and (b) captures the first and last popped values and verifies first = 999 and last = 0 (descending pop order). Adds two assignments per loop iteration but catches both `make_heap` and `pop_heap` regressions.

File: benchmarks/sort/heap_ops.das [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+                sort_boost::push_heap(buf)
+            } elif (x < buf[0]) {
+                sort_boost::pop_heap(buf)
+                buf[length(buf) - 1] = x
+                sort_boost::push_heap(buf)
+            }
+        }
+        if (length(buf) != cap) {
+            b->failNow()
+        }
```
[the remark this answers is not in the document]
Comment (borisbat) [reply benchmarks/sort/heap_ops.das:55]: Fixed in 0c1813de4 — `bounded_topn` now checks `buf[0] != expected_sorted_value_at(n, cap - 1)`. After the streaming loop the heap root is the max of the kept set, which must equal the cap-th smallest source value (cap=100, n=100000: kth value = 0; cap=1000, n=100000: kth value = 9). Catches no-op `push_heap` / `pop_heap`.

File: daslib/linq.das [hunk 11]
```diff
@@ -448,6 +449,64 @@ def order_by_descending_to_array(var a : iterator<auto(TT)>; key) : array<TT -co
...
+// das_pop_heap_r (byte-pointer heap on the any-cblock path).
+// ============================================================================
+
+def top_n_by(arr : array<auto(TT)>; n : int; key) : array<TT -const -&> {
+    //! Returns the ``n`` smallest elements of ``arr`` by ``key(element)``,
+    //! sorted ascending. ``n <= 0`` or empty ``arr`` yields an empty result.
+    var buf : array<TT -const -&>
+    if (n <= 0 || empty(arr)) return <- buf
+    let take_count = min(n, length(arr))
+    buf |> reserve(take_count)
```
[the remark this answers is not in the document]
Comment (borisbat) [reply daslib/linq.das]: Fixed in 0c1813de4 — `reserve(length(arr))` instead of `reserve(take_count)`. Loop pushes every element, so the original `reserve(min(n, length))` was indeed too small whenever `n << length(arr)` and caused repeated growths. We still `resize(take_count)` after the `partial_sort` to return only the N smallest.

**Review by borisbat — COMMENTED** [review]
@copilot please re-review — 0c1813de4 addresses your inline feedback (9 of 11 accepts, 1 reject on the .rst half-open-range claim, 1 confirmed pre-existing master fix bundled in). New benchmark numbers attached below.

**Comment by borisbat** [comment]
## Benchmark numbers — `benchmarks/sql/sort_take.das` (n = 100000, INTERP)

| Column | ns/op | vs SQL (m1) | vs `order_by + take` (m3) |
|---|---:|---:|---:|
| m1 — SQL `ORDER BY price LIMIT 10` | 37 | 1.0× | 0.017× |
| m3 — array `order_by(price) \|> take(10)` | 2175 | 58.8× | 1.0× |
| m3f_old — frozen baseline `_old_fold` | 2205 | 59.6× | 1.01× |
| m3f — current `_fold` | 2195 | 59.3× | 1.01× |
| **m3_topn_array (new) — `top_n_by(arr, 10, key)`** | **442** | **11.9×** | **0.20× (5× faster)** |
| **m3_topn_iter (new) — `top_n_by(arr.to_sequence(), 10, key)`** | **83** | **2.2×** | **0.038× (26× faster)** |

Phase 0 story:
- SQL ORDER BY ... LIMIT 10 with a real index runs at 37 ns/op
- The existing LINQ `order_by \|> take` is a full O(M log M) sort then slice — **59× slower than SQL** at M=100K
- New `top_n_by(array, N, key)` uses `partial_sort` underneath (introselect + smoothsort tail). O(M log N) — **5× faster than order_by+take, 12× off SQL**
- New `top_n_by(iterator, N, key)` uses a bounded heap of size N during the scan (max N+1 elements ever resident). O(M log N) with much smaller constants — **26× faster than order_by+take, only 2.2× off SQL**

The bounded-heap streaming form essentially closes the gap; the remaining 2.2× is the cost of LINQ blocks/keys + the lambda invoke per element vs SQLite's tight C loop.

This is exactly the primitive PR B (`BufferTopN` emit mode in `linq_fold`) needs — when the splicer detects `[where][select]* |> order_by(...) |> {take(N), first}`, it can pick the bounded-heap-during-scan strategy without any user-side code change.

File: daslib/sort_boost.das [hunk 12]
```diff
@@ -6,15 +6,287 @@ options strict_smart_pointers = true
...
-//! Extended sorting support.
+//! Extended sorting support: partial_sort, nth_element, heap operations, and the
+//! ``qsort`` family of dispatcher macros that route to the right ``__builtin_*``
+//! variant based on container type.
 //!
-//! Provides the ``qsort`` macro that dispatches to `sort` for built-in arrays
-//! and ``sort(temp_array(...))`` for handled vector types.
+//! Phase 0 wrappers cover ``array<T>`` only; for fixed-size dim or handled
+//! vector types use the matching ``q*`` macro (see ``qsort``-style dispatch
+//! macros at the bottom of this module).
```
[the remark this answers is not in the document]
Comment (borisbat) [reply daslib/sort_boost.das]: Fixed in 012c4159d — module preamble rewritten to say "Phase 0 wrappers cover `array<T>` directly; for handled vector types use the matching `q*` macro, which wraps the input in `temp_array` before dispatching." No more dim mention.

File: daslib/sort_boost.das [hunk 13]
```diff
@@ -45,3 +317,78 @@ class QsortMacro : AstCallMacro {
...
+    } else {
+        macro_error(prog, expr.at, "can only {opName} array or handled vector")
+        return null
+    }
+}
+
+[call_macro(name="qpartial_sort")]
+class QPartialSortMacro : AstCallMacro {
+    //! Implements ``qpartial_sort(value, n, block)`` — dispatches to
+    //! ``partial_sort`` over array/dim or ``partial_sort(temp_array(value), …)``
```
[the remark this answers is not in the document]
Comment (borisbat) [reply daslib/sort_boost.das]: Fixed in 012c4159d — all 5 q* macro doc-comments now say "for arrays and handled vectors" without mentioning dim.

File: benchmarks/sort/_common.das [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+    }
+    return <- arr
+}
+
+def public expected_sorted_value_at(n : int; i : int) : int {
+    //! For input ``((i + 1) * 37) % 1000`` with ``n`` >= 1000, all 1000
+    //! residues are hit uniformly (``n / 1000`` copies each), so sorted
+    //! ``arr[i] = (i * 1000) / n``. Used by benchmarks for self-validating
+    //! post-sort checks.
+    return (i * 1000) / n
```
[the remark this answers is not in the document]
Comment (borisbat) [reply benchmarks/sort/_common.das]: Fixed in 012c4159d — docstring tightened to "for n a multiple of 1000" and explicitly notes the helper is not safe for arbitrary n. Formula is exact at the n values we actually exercise (10000, 100000); kept as-is rather than generalized since the bench fixtures only use multiples of 1000.

File: daslib/sort_boost.das [hunk 15]
```diff
@@ -45,3 +317,78 @@ class QsortMacro : AstCallMacro {
     }
 }
 
+// Internal helper: build an ``opName(value [, n], block)`` call from the
+// macro's call expression, wrapping ``value`` in temp_array for handled vectors.
+def private build_dispatch_call(prog : ProgramPtr; var expr : ExprCallMacro?; opName : string; takesN : bool) : ExpressionPtr {
+    let expectedArgs = takesN ? 3 : 2
+    let argShape = takesN ? "(value, n, block)" : "(value, block)"
+    macro_verify(expr.arguments |> length == expectedArgs, prog, expr.at, "expecting {opName}{argShape}")
```
[the remark this answers is not in the document]
Comment (borisbat) [reply daslib/sort_boost.das]: Fixed in 012c4159d — `build_dispatch_call` now takes a separate `macroName` parameter alongside `opName`. The wrong-arg-count error reports `expecting qpartial_sort(value, n, block)` (macro name), and the unsupported-source error reports `qpartial_sort can only operate on array or handled vector`. All 5 q* macros updated to pass both names.

File: src/builtin/das_qsort_r.h [hunk 16]
```diff
@@ -230,4 +232,148 @@ inline void das_qsort_r(void *base, uint32_t nel, uint32_t width, Compare cmp)
...
+static inline void byte_swap(void *pa, void *pb, uint32_t width)
+{
+    unsigned char tmp[256];
+    unsigned char *a = (unsigned char *)pa;
+    unsigned char *b = (unsigned char *)pb;
+    while (width) {
+        uint32_t chunk = uint32_t(sizeof(tmp)) < width ? uint32_t(sizeof(tmp)) : width;
+        memcpy(tmp, a, chunk);
+        memcpy(a, b, chunk);
+        memcpy(b, tmp, chunk);
```
[the remark this answers is not in the document]
Comment (borisbat) [reply src/builtin/das_qsort_r.h:257]: Fixed in 012c4159d — `#include <cstring>` added to `das_qsort_r.h`. Same root cause as the cstdint fix in the prior round: Mac SDK pulls memcpy in transitively; Linux/MSVC do not. This was the CI failure across 12 builds.

File: daslib/sort_boost.das [hunk 17]
```diff
@@ -45,3 +317,78 @@ class QsortMacro : AstCallMacro {
...
+    }
+}
+
+[call_macro(name="qpartial_sort")]
+class QPartialSortMacro : AstCallMacro {
+    //! Implements ``qpartial_sort(value, n, block)`` — dispatches to
+    //! ``partial_sort`` over array/dim or ``partial_sort(temp_array(value), …)``
+    //! for handled vector types.
+    def override visit(prog : ProgramPtr; mod : Module?; var expr : ExprCallMacro?) : ExpressionPtr {
+        return <- build_dispatch_call(prog, expr, "partial_sort", true)
```
[the remark this answers is not in the document]
Comment (borisbat) [reply daslib/sort_boost.das]: Fixed in 012c4159d — added `test_q_macros` block to `tests/linq/test_linq_sorting.das` with 5 subtests covering `qpartial_sort` / `qnth_element` / `qmake_heap` / `qpush_heap` / `qpop_heap` dispatch on `array<int>`. 59/59 pass (was 53). Note: the pre-existing `qsort` macro also has no test in the tree; kept that out of scope for Phase 0 to avoid scope creep

File: daslib/linq.das [hunk 18]
```diff
@@ -448,6 +449,64 @@ def order_by_descending_to_array(var a : iterator<auto(TT)>; key) : array<TT -co
     return <- arr
 }
 
+// ============================================================================
+// top_n / top_n_by — return the N smallest elements in ascending order.
+//
+// Array source: partial_sort on a clone, then resize to N. O(M log N).
+// Iterator source: bounded heap of size N maintained during scan, then sort.
+// Iteratively O(M log N), max N+1 elements ever resident.
```
[the remark this answers is not in the document]
Comment (borisbat) [reply daslib/linq.das]: Fixed in 012c4159d — comment now says "at most `n` elements resident". You are right — the pop_heap, overwrite-last-slot, push_heap sequence keeps the buffer length at n the entire time; n+1 was wrong.

File: doc/source/reference/tutorials/33_algorithm.rst [hunk 19]
```diff
@@ -49,6 +53,135 @@ Custom comparators let you search in non-default order:
...
+.. code-block:: das
+
+    require daslib/linq
+
+    let arr <- [10, 20, 5, 8, 30, 15, 2, 25]
+    let smallest3 <- top_n(arr, 3)
+    // [2, 5, 8]
+
+For iterator sources, ``top_n`` maintains a bounded heap of size ``N`` during
+the scan — at most ``N+1`` elements are ever resident:
```
[the remark this answers is not in the document]
Comment (borisbat) [reply doc/source/reference/tutorials/33_algorithm.rst]: Fixed in 012c4159d — tutorial RST line now says "at most `N` elements are ever resident". Same fix as linq.das.

File: tutorials/language/33_algorithm.das [hunk 20]
```diff
@@ -51,6 +66,122 @@ def binary_search_demo() {
...
+    print("\n=== top_n / top_n_by ===\n")
+
+    // top_n returns the N smallest elements as a sorted ascending array.
+    // For array sources it uses partial_sort internally (O(M log N)).
+    let arr <- [10, 20, 5, 8, 30, 15, 2, 25]
+    let smallest3 <- top_n(arr, 3)
+    print("top_n(arr, 3): {smallest3}\n")  // [2, 5, 8]
+
+    // For iterator sources, top_n uses a bounded heap of size N during the
+    // scan — at most N+1 elements ever resident, O(M log N) total.
```
[the remark this answers is not in the document]
Comment (borisbat) [reply tutorials/language/33_algorithm.das]: Fixed in f74c26325 — `tutorials/language/33_algorithm.das:166` now says "at most N elements ever resident". (I missed this one in 012c4159d; matches the linq.das and 33_algorithm.rst fixes.)

File: benchmarks/sql/sort_take.das [hunk 21]
```diff
@@ -2,12 +2,22 @@ options gen2
...
 
 let TAKE_N = 10
 
 // _order_by(_.price) |> take(N) — top-N (here bottom-N since ascending) pattern.
 // SQL: ORDER BY price LIMIT N — engine emits partial-sort optimization when possible.
 // Array variants must fully sort before indexing the first N.
+//
+// The m3_topn_* columns measure the new ``linq.top_n_by`` family: array source
+// uses ``partial_sort`` under the hood (O(M log N)), iterator source uses a
+// bounded heap of size N during the scan (max N+1 elements resident).
```
[the remark this answers is not in the document]
Comment (borisbat) [reply benchmarks/sql/sort_take.das]: Fixed in f74c26325 — `benchmarks/sql/sort_take.das:16` now says "max N elements resident". Same root cause as the linq.das comment. Also self-spotted and fixed one more in `daslib/linq.das:457` (section header above top_n) that I missed in 012c4159d.

File: daslib/linq.das [hunk 22]
```diff
@@ -448,6 +449,64 @@ def order_by_descending_to_array(var a : iterator<auto(TT)>; key) : array<TT -co
...
+    return <- buf
+}
+
+def top_n_by(var a : iterator<auto(TT)>; n : int; key) : array<TT -const -&> {
+    //! Returns the ``n`` smallest elements of an iterator by ``key(element)``,
+    //! sorted ascending. Uses a bounded heap of size ``n`` during the scan;
+    //! at most ``n`` elements resident.
+    var buf : array<TT -const -&>
+    if (n <= 0) return <- buf
+    buf |> reserve(n)
```
[the remark this answers is not in the document]
Comment (borisbat) [reply daslib/linq.das]: Fixed in 6c065d15e — dropped `buf |> reserve(n)`. daslang `array<T>::push` already grows geometrically, so the O(log n) doublings during fill amortize to O(1) per push — no asymptotic change but the worst-case foot-gun (n=1M against a short iterator) is gone. Added a 3-line comment explaining the choice. The array-source overload keeps `reserve(length(arr))` because cardinality is known there. 59/59 tests still pass.

File: src/builtin/module_builtin_runtime_sort.cpp [hunk 23]
```diff
@@ -239,6 +511,86 @@ namespace das
         addExtern<DAS_BIND_FUN(builtin_sort_dim_any_ref_cblock)>(*this, lib, "__builtin_sort_dim_any_ref_cblock",
             SideEffects::modifyArgumentAndExternal, "builtin_sort_dim_any_ref_cblock_T")
                 ->args({"array","stride","length","block","context","line"})->setAotTemplate()->setAnyTemplate();
+
+        // ==================================================================
+        // Phase 0: partial_sort / nth_element / make_heap / push_heap / pop_heap.
+        // Same 19-type matrix as sort (6 numeric + 13 vector). String is omitted
+        // for now — strudel/sort_string is a special path with strcmp comparator;
```
[the remark this answers is not in the document]
Comment (borisbat) [reply src/builtin/module_builtin_runtime_sort.cpp]: Fixed in 6c065d15e — comment now says `__builtin_sort_string` correctly. You are right, "strudel" was completely off (that is the music module, unrelated to strings). Mistyped during the original drop.
